### PR TITLE
Pass manager options at registration time

### PR DESCRIPTION
**Pull request by wswoodruff**
- Feedback welcome

- Set multiple: true, as it's required to pass plugin options
- Register vision via `server.register({ plugin: Vision, options: { ... });` where:
    - options is a manager config, is synonymous with `server.views`
- Effectively, the only thing this adds is the ability to use

        await server.register({
            plugin: Vision,
            options: {/* view manager options */}
        });

        // which will have the same effect as

        await server.register(Vision);
        server.views({/* view manager options */});

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -122,6 +128,37 @@ internals.handler = function (route, options) {
...
+
+internals.nearestState = function (realm) {
+
+    if (realm.plugins.vision) {
+        return realm.plugins.vision;
+    }
+
+    const parent = realm.parent;
+
+    return internals.nearestState(parent);
```
Comment (devinivy) [inline lib/index.js]: Perhaps (?) not a practical concern here, but beware recursion in javascript.  There's still no such thing as tail recursion in node, so each time you recurse the call stack gets longer (and it has a max allowed size).

File: lib/index.js [hunk 2]
```diff
@@ -122,6 +128,37 @@ internals.handler = function (route, options) {
...
+};
+
+
+internals.getRootRealm = (realm) => {
+
+    let rootRealm = realm;
```
Comment (devinivy) [inline lib/index.js]: I think `rootRealm` is a slightly odd variable name here– it's not the root realm until it's returned (and not being used as a variable anymore).

File: lib/index.js [hunk 3]
```diff
@@ -11,89 +11,95 @@ const Manager = require('./manager');
...
 
-            const realm = (this.realm.plugins.vision || this.request.server.realm.plugins.vision || {});
-            Hoek.assert(realm.manager, 'Cannot render view without a views manager configured');
-            return this.response(realm.manager._response(template, context, options, this.request));
-        });
+internals.assignManager = function (options) {
+
+    Joi.assert(options, internals.schema.manager, 'Invalid manager options');
+
+    const realm = this.realm.parent || this.realm;
```
Comment (devinivy) [inline lib/index.js]: Wait, when `server.views()` is called inside a plugin wont this set the manager on the plugin's parent?  It would be hard to see this in a test when `nearestState()` comes into play.  But imagine `server1` registers a plugin with `server2`, and `server2` register a plugin with `server3`.  Now `server3` calls `server3.views()`– I think this will cause `server2` and `server3` to both have the same manager.

File: lib/index.js (hunk 1 above)
Comment (devinivy) [reply lib/index.js]: (And this can always be expressed as a while loop!)

File: lib/index.js [hunk 4]
```diff
@@ -11,89 +11,95 @@ const Manager = require('./manager');
...
 
-        server.decorate('toolkit', 'view', function (template, context, options) {
 
-            const realm = (this.realm.plugins.vision || this.request.server.realm.plugins.vision || {});
-            Hoek.assert(realm.manager, 'Cannot render view without a views manager configured');
-            return this.response(realm.manager._response(template, context, options, this.request));
-        });
+internals.assignManager = function (options) {
+
+    Joi.assert(options, internals.schema.manager, 'Invalid manager options');
```
Comment (devinivy) [inline lib/index.js]: Doesn't the manager make this assertion on its own?

File: test/index.js [hunk 5]
```diff
@@ -437,15 +437,17 @@ describe('views()', () => {
...
 
-            register: function (server, options) {
+            register: async function (server, options) {
 
                 server.path(__dirname);
 
                 const views = {
                     engines: { 'html': Handlebars },
-                    path: './templates/plugin'
+                    path: __dirname + '/templates/plugin'
```
Comment (devinivy) [inline test/index.js]: Why this change?

File: test/index.js [hunk 6]
```diff
@@ -578,27 +580,215 @@ describe('views()', () => {
...
-    it('only registers once', async () => {
+    it('registers multiple times', async () => {
 
         const one = {
             name: 'one',
-
             register: function (server, options) {
 
-                return server.register(Vision);
+                server.register({
```
Comment (jedireza) [inline test/index.js]: Should we be using awaiting the return of this? (seen later in other call sites as well)

```js
await server.register(...
```

File: test/index.js [hunk 7]
```diff
@@ -578,27 +580,215 @@ describe('views()', () => {
...
+                    plugin: Vision,
+                    options: {
+                        engines: { 'html': Handlebars },
+                        relativeTo: Path.join(__dirname, '/templates'),
+                        path: 'plugin'
+                    }
+                });
+
+                return server.route({
```
Comment (jedireza) [inline test/index.js]: I'm curious why you're using `return` here. I don't see a reference in the docs about needing to return anything from a plugin's `register` function.  (seen later in other call sites as well)

**Comment by wswoodruff** [comment]
Thanks so much for the review guys!

File: test/index.js (hunk 6 above)
Comment (wswoodruff) [reply test/index.js]: I've asked around about this, "what's the difference between `return await [Promise]` and `return [Promise]`? And `ljharb` in the `Node.js` Slack community said they're identical in what they do, only returning an await causes an unnecessary wrapper. Do you know if there's a difference?

File: test/index.js (hunk 5 above)
Comment (wswoodruff) [reply test/index.js]: This might've slipped through when trying to write tests based off of others. I can remove it!

File: lib/index.js (hunk 2 above)
Comment (wswoodruff) [reply lib/index.js]: Agree!

File: lib/index.js (hunk 1 above)
Comment (wswoodruff) [reply lib/index.js]: I was having trouble getting code coverage with the while loop Eran had here, because `parent` would never evaluate to false, since the root server _always_ has vision state. The root server always has vision state because of the `{ setup: true }` introduced in the `register` func.

Here's what's currently in vision - the loop Eran wrote:
```
internals.realm = function (server) {

    if (server.realm.plugins.vision) {
        return server.realm.plugins.vision;
    }

    let parent = server.realm.parent;
    while (parent) {
        if (parent.plugins.vision) {
            return parent.plugins.vision;
        }

        parent = parent.parent;
    }

    return {};
};
```

As soon as the root server goes in this loop, it is returned as having a realm because setup.true will always be there. So `parent = parent.parent` will never run, which would cause it to evaluate to false in the while loop, and return `{}`

Comment (wswoodruff) [reply lib/index.js]: This could possibly be rewritten as `nearestManager`, which would fix the issue of returning just because a realm has state in vision

File: lib/index.js (hunk 3 above)
Comment (wswoodruff) [reply lib/index.js]: I think you're probably right, I'll definitely go back and ensure this acts correctly in tests

File: test/index.js (hunk 6 above)
Comment (jedireza) [reply test/index.js]: I see. That's good to know.

> Do you know if there's a difference?

Nope. I'm new to async/await in general.

Comment (devinivy) [reply test/index.js]: I find it's generally easier to remove cognitive load by simply awaiting anything async– only think about the underlying promises when really necessary.  For what it's worth, in a lot of hapi code you'll find `return await somePromise`.  Unless perf is terribly important, that seems most consistent.

Comment (wswoodruff) [reply test/index.js]: Good to know!!

**Comment by thebinarypenguin** [comment]
I'm excited about the ability to pass view manager options at registration time, but I'm curious what does that have to do with multiple registrations?

**Comment by wswoodruff** [comment]
https://github.com/hapijs/hapi/blob/master/API.md#-await-serverregisterplugins-options
mentions that `once: true` ... 'Cannot be used with plugin options.'

**Comment by thebinarypenguin** [comment]
@wswoodruff Gotcha. Thanks for the plugin, and preemptive thanks for this new feature. :tada:

File: lib/index.js [hunk 8]
```diff
@@ -12,88 +12,101 @@ const Manager = require('./manager');
...
-        }
+    // This fingerprint is set when using the 'onRequest' server ext
+    if (this.request.route.fingerprint === '/#') {
 
-        parent = parent.parent;
+        // 'this' is the server that registered the extension
+        realm = this.realm;
+    }
+    else {
+        realm = this.request.route.realm;
```
Comment (wswoodruff) [inline lib/index.js]: I'm curious what others think of using the fingerprint `/#` for detecting when using the 'onRequest' server extension.

File: lib/index.js [hunk 9]
```diff
@@ -12,88 +12,101 @@ const Manager = require('./manager');
...
-    }
+    let realm;
 
-    let parent = server.realm.parent;
-    while (parent) {
-        if (parent.plugins.vision) {
-            return parent.plugins.vision;
-        }
+    // This fingerprint is set when using the 'onRequest' server ext
+    if (this.request.route.fingerprint === '/#') {
```
Comment (wswoodruff) [inline lib/index.js]: I'm curious what others think of using the fingerprint `/#` for detecting when using the 'onRequest' server extension.

File: lib/index.js [hunk 10]
```diff
@@ -12,88 +12,101 @@ const Manager = require('./manager');
...
 
-        parent = parent.parent;
+        // 'this' is the plugin/server that registered the extension
```
Comment (devinivy) [inline lib/index.js]: `this` should be the toolkit itself.  If `h.view()` is called in an extension then `h.realm` will be the realm in which the extension was created.

I'm pretty sure you want to always use `h.realm` and not `h.request.route.realm` here.  Imagine writing a plugin that has its own manager and registers some extensions (but no routes).  When calling `h.view()` in one of those extensions I think you would expect to know only about your manager/templates, and not about other managers/templates registered in a different plugin (i.e. a separate plugin that registers routes).

Comment (wswoodruff) [reply lib/index.js]: Hm, now I can't remember why I wrote it grabbing the realm from the route.

**Comment by wswoodruff** [comment]
Thanks for the reviews guys! I'm hoping to merge this to master soon after the holidays, gonna let it marinate again for a little bit, not a whole month this time tho loL

**Comment by wswoodruff** [comment]
It's time! =P
